micro/goal.go

```diff
@@ -41,11 +48,8 @@ func (s Substitutions) String() string {
 	return l[1 : len(l)-1]
 }
 
-// GoalFn is a function that takes a state and returns a stream of states.
-type GoalFn func(*State) StreamOfStates
-
-// Goal is a function that returns a GoalFn. Used for lazy evaluation
```

Comment:
I didn't know that we didn't need lazy evaluation, could you explain a bit how you figured this out?

Reply:
Lazy evaluation is now fully handled by the stream implementation, instead of bleeding over into goal implementation. This is reverting back to your original implementation of Goal